File: finalists/Ximo/challenge_10/10.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function, division
input=raw_input
import sys

if __name__ == '__main__':
    for line in sys.stdin:
        ops = []
        for o in line.split():
            if o.isdigit():
                ops.append(int(o))
            elif o == 'mirror':
                ops[-1] *= -1
            elif o == 'breadandfish':
                ops.append(ops[-1])
            elif o == '#':
                b = ops.pop()
                a = ops.pop()
                ops.append(a*b)
            elif o == 'conquer':
                b = ops.pop()
                a = ops.pop()
                ops.append(a%b)
            elif o == 'dance':
                ops[-1], ops[-2] = ops[-2], ops[-1]
            elif o == '.':
                print(ops[-1])
            elif o == '$':
                b = ops.pop()
                a = ops.pop()
                ops.append(a-b)
            elif o == 'fire':
                b = ops.pop()
                a = ops.pop()
                ops.append(max(a,b))
            elif o == '@':
                b = ops.pop()
                a = ops.pop()
                ops.append(a+b)
            elif o == '&':                
                b = ops.pop()
                a = ops.pop()
                ops.append(a//b)
            else:
                print('o=', o)
                break
 

File: finalists/Carlos/challenge_10/10.pl
#!/usr/bin/perl

use strict;

my $line;

# save input data, just for review
open F, ">input_data.txt";

while ( ($line = <>) )
{
  print F $line;
  chomp $line;

  my @stack = ();

  # operators
  my @ops = split(/\s+/, $line);

  my $reg;
  my $reg_temp;

  my $i;

  # Reverse Polish notation
  for ($i = 0; $i < scalar @ops; $i++)
  {
    $reg = $ops[$i];

    if ($reg eq "mirror")
    {
      $reg = pop(@stack);
      $reg = -$reg;
      push(@stack, $reg);
    }
    elsif ($reg eq "breadandfish")
    {
      $reg = pop(@stack);
      push(@stack, $reg);
      push(@stack, $reg);
    }
    elsif ($reg eq "#")
    {
      $reg = pop(@stack) * pop(@stack);
      push(@stack, $reg);
    }
    elsif ($reg eq "fire")
    {
      $reg = pop(@stack);
    }
    elsif ($reg eq '$')
    {
      $reg = pop(@stack);
      $reg = pop(@stack) - $reg;
      push(@stack, $reg);
    }
    elsif ($reg eq "dance")
    {
      $reg = pop(@stack);
      $reg_temp = pop(@stack);
      push(@stack, $reg);
      push(@stack, $reg_temp);
    }
    elsif ($reg eq "conquer")
    {
      $reg = pop(@stack);
      $reg_temp = pop(@stack);
      $reg = $reg_temp % $reg;
      push(@stack, $reg);
    }
    elsif ($reg eq "&")
    {
      $reg = pop(@stack);
      $reg = pop(@stack) / $reg;
      push(@stack, $reg);
    }
    elsif ($reg eq "@")
    {
      $reg = pop(@stack) + pop(@stack);
      push(@stack, $reg);
    }
    elsif ($reg eq ".")
    {
      $reg = pop(@stack);
      print "$reg\n";
    }
    elsif ($reg ne "")
    {
      push(@stack, $reg);
    }
  }
}


close F;

